icicle-source/src/Icicle/Source/ToCore/ToCore.hs

```diff
@@ -593,3 +602,99 @@ convertReduce xx
+  let sndX  t u = CE.xPrim . C.PrimMinimal . Min.PrimPair     $ Min.PrimPairSnd    t u
+  let someX t   = CE.xPrim . C.PrimMinimal . Min.PrimConst    $ Min.PrimConstSome  t
+  let eqX   t   = CE.xPrim . C.PrimMinimal $ Min.PrimRelation   Min.PrimRelationEq t
+
+  let nubKons t'acc n'input xx = do
+        n'acc'old <- lift fresh
+        n'acc'new <- lift fresh
+        n'key'old <- lift fresh
+        n'key'new <- lift fresh
+
+        xx' <- lift $ CE.subst1 () n'input (CE.xVar n'acc'old) xx
+        let a  =                    CE.makeApps () (sndX  t'key t'acc) [CE.xVar n'input]
+        let x' = CE.xLet n'key'old (CE.makeApps () (fstX  t'key t'acc) [CE.xVar n'input])
+               $ CE.xLet n'acc'old a
+               $ CE.xLet n'key'new (CE.makeApps () (someX t'k        ) [k'             ])
+               $ CE.xLet n'acc'new xx'
+               $ CE.makeApps ()
+                   ( CE.xPrim (C.PrimFold C.PrimFoldBool (T.PairT t'key t'acc)) )
+                   [ CE.xVar n'input
+                   , CE.makeApps () (pairX t'key t'acc) [ CE.xVar n'key'new, CE.xVar n'acc'new ]
+                   , CE.makeApps () (eqX   t'key      ) [ CE.xVar n'key'old, CE.xVar n'key'new ] ]
+
+        pure ( x', Map.singleton n'input a )
+
+  let nubInit t'acc xx
+        = CE.makeApps () (pairX t'key t'acc) [ CE.xValue t'key VNone, xx ]
+
+  -- Substitute the name of the stream input in the body of this stream and all downstreams,
+  -- since the stream input should now refer to the key paired with the original input.
+  let substStream subs (C.SFold f t ini kons)
+        =   C.SFold f t
+        <$> CE.subst () subs ini
+        <*> CE.subst () subs kons
+      substStream subs (C.SFilter f ss)
+        =   C.SFilter f
+        <$> mapM (substStream subs) ss
+
+  let substThen f (s : ss) = do
+        (s', substs)   <- f [s]
+        (ss', substs') <- f =<< lift (mapM (substStream substs) ss)
+        pure (s' <> ss', substs <> substs')
+      substThen _ []
+        = pure ([], Map.empty)
+
+  -- Prefix each streams with the nub expression.
+  let nub (C.SFold n t ini kons : downstreams) = do
+        (x', sub)    <- nubKons t n kons
+        downstreams' <- lift $ mapM (substStream sub) downstreams
+        (ss, subs)   <- nub downstreams'
+        let s = C.SFold n (T.PairT t'key t) (nubInit t ini) x'
+        pure (s : ss, sub <> subs)
+
+      nub (C.SFilter x fs : downstreams) = do
+        (fs', sub)    <- substThen nub fs
+        downstreams'  <- lift $ mapM (substStream sub) downstreams
+        (ss, subs)    <- nub downstreams'
+        let s = C.SFilter x fs'
+        pure (s : ss, sub <> subs)
+
+      nub [] = pure ([], Map.empty)
+
+  (streams', subs) <- hoist runFreshIdentity . nub
+                    . streams $ core
+  postcomps'       <- lift . runFreshIdentity
+                    . mapM (\(n, x) -> (n,) <$> CE.subst () subs x)
+                    . postcomps $ core
```

Comment:
I didn't realise you could do it without modifying convertQuery/convertReduce, so this is really cool